packages/core/src/electron-main/electron-application.ts

```diff
@@ -0,0 +1,466 @@
+        // Set the electron version for both the dev and the production mode. (https://github.com/eclipse-theia/theia/issues/3254)
+        // Otherwise, the forked backend processes will not know that they're serving the electron frontend.
+        process.env.THEIA_ELECTRON_VERSION = process.versions.electron;
+        if (noBackendFork) {
+            process.env[ElectronSecurityToken] = JSON.stringify(this.electronSecurityToken);
+            // The backend server main file is supposed to export a promise resolving with the port used by the http(s) server.
+            const address: AddressInfo = await require(this.globals.THEIA_BACKEND_MAIN_PATH);
+            return address.port;
+        } else {
+            const backendProcess = fork(
+                this.globals.THEIA_BACKEND_MAIN_PATH,
+                this.processArgv.getProcessArgvWithoutBin(),
+                await this.getForkOptions(),
+            );
+            return new Promise((resolve, reject) => {
+                // The backend server main file is also supposed to send the resolved http(s) server port via IPC.
+                backendProcess.on('message', (address: AddressInfo) => {
+                    resolve(address.port);
+                });
+                backendProcess.on('error', error => {
+                    reject(error);
+                });
+                app.on('quit', () => {
+                    // If we forked the process for the clusters, we need to manually terminate it.
+                    // See: https://github.com/eclipse-theia/theia/issues/835
+                    process.kill(backendProcess.pid);
+                });
+            });
+        }
+    }
+
+    protected async getForkOptions(): Promise<ForkOptions> {
+        return {
+            env: {
+                ...process.env,
+                [ElectronSecurityToken]: JSON.stringify(this.electronSecurityToken),
+            },
+        };
+    }
+
+    protected async attachElectronSecurityToken(port: number): Promise<void> {
+        await new Promise((resolve, reject) => {
+            session.defaultSession!.cookies.set({
+                url: `http://localhost:${port}`,
+                name: ElectronSecurityToken,
+                value: JSON.stringify(this.electronSecurityToken),
+                httpOnly: true,
+            }, error => error ? reject(error) : resolve());
+        });
+    }
+
+    protected hookApplicationEvents(): void {
+        app.on('will-quit', this.onWillQuit.bind(this));
+        app.on('second-instance', this.onSecondInstance.bind(this));
+        app.on('window-all-closed', this.onWindowAllClosed.bind(this));
+    }
+
+    protected onWillQuit(event: ElectronEvent): void {
+        this.stopContributions();
+    }
+
+    protected async onSecondInstance(event: ElectronEvent, argv: string[], cwd: string): Promise<void> {
+        await this.launch({ argv: this.processArgv.getProcessArgvWithoutBin(argv), cwd, secondInstance: true });
```

Comment:
I'm not sure this is the correct way to handle the [second-instance](https://www.electronjs.org/docs/api/app#event-second-instance) event, although I haven't tested this PR in a packaged app.

On master, the generated electron-main file checks for single-instance mode, and focuses the current window. What is the reasoning for this change?

Reply:
> I'm not sure this is the correct way to handle the [second-instance](https://www.electronjs.org/docs/api/app#event-second-instance) event, although I haven't tested this PR in a packaged app.

@marechal-p, can you please help with this? Thank you!

Reply:
> On master, the generated electron-main file checks for single-instance mode, and focuses the current window. What is the reasoning for this change?

I thought that it would have been better to support a use case where a user does: `app path/to/workspace` and it opens a window with that workspace opened.

Fortunately this behavior is now overridable, wdyt @mcgordonite ?

Reply:
More importantly, the old behavior had the concept of some kind of "main window" which would disregard other windows opened. What if you open different workspaces and close the first one? That felt weird to me so I removed this concept of "main window". Now you just have windows and react when everything is closed.

Reply:
I guess after this patch, one can add logic to go through each window on the second-instance event and decide what to do, e.g. find if a window is already opened with said workspace and focus this instead of reopening a new window.

Reply:
Spamming with comments, sorry, to to further explain the way I see this change: To me `second-instance` meant single electron-main process but multiple browser windows.

Reply:
I see, so we want to open a new `BrowserWindow` instead of launching a whole new electron main process when the user runs the app again?

In that case, I think we always need to `requestSingleInstanceLock` otherwise the `second-instance` event won't fire. I suggest moving the handling for the `isSingleInstance` configuration into `ElectronApplication` if we can. 

I'm just investigating if that is possible by building some installers from this PR.

Reply:
I created a PR with a suggestion here: https://github.com/eclipse-theia/theia/pull/8130.